Proxyclient asm toolchain refactor

alxdrl opened the pull request:
Proxyclient toolchain is slightly broken with recent macOS/brew systems with llvm / lld packages.

In the process of fixing it I tried to untangle the logic but I quickly over-engineered the thing :
- kept existing interface for environment variables override and tool variable names
- tried to isolate platform/path-dependent info in the new classes
- tried to decouple things for ease of testing
- not sure about LDFLAGS removal...
- not sure about handling of CFLAGS ...

This is only lightly tested on macOS with brew (passes the assert test in asm.py) and absolutely not tested on Linux/OpenBSD.

@asdfugil looking at your tree I understand you also work from macOS host and might be interested in this.

svenpeter42 commented:
I'm sometimes working from a macOS host as well and have hacked around the issues I occasionally ran into so far. Happy to see someone over-engineering this instead ;)

Only took a brief glance but this looks reasonable so far.

alxdrl commented:
@svenpeter42 thanks for the feedback, will try to give this some actual test and resubmit once I'm confident this does not break.

asdfugil commented:
I have been using these patches on Linux and macOS for about a month, and they are working fine so far.

jvoisin commented on proxyclient/m1n1/asm.py:
```diff
@@ -72,10 +37,10 @@ def compile(self, source):
-        self._call(LD, f"{self.LDFLAGS} --Ttext={self.addr:#x} -o {self.elffile} {self.ofile}")
-        self._call(OBJCOPY, f"-j.text -O binary {self.elffile} {self.bfile}")
-        self._call(NM, f"{self.elffile} > {self.nfile}")
+        self._call(self.toolchain.CC, f" -c -o {self.ofile} {self.sfile}")
+        self._call(self.toolchain.LD, f" --Ttext={self.addr:#x} -o {self.elffile} {self.ofile}")
+        self._call(self.toolchain.OBJCOPY, f" -j.text -O binary {self.elffile} {self.bfile}")
```
```suggestion
        self._call(self.toolchain.CC, f"-c -o {self.ofile} {self.sfile}")
        self._call(self.toolchain.LD, f"--Ttext={self.addr:#x} -o {self.elffile} {self.ofile}")
        self._call(self.toolchain.OBJCOPY, f"-j.text -O binary {self.elffile} {self.bfile}")
```

no need for the odd leading space.

svenpeter42 commented:
Nice, even tests now! Took another quick glance and this still looks very good to me. I'll take a closer look and give it a test on macOS once you'll mark it as ready.

alxdrl commented:
- added requirements.in/requirements.txt
- added test workflow
- skipped tests when relevant
- factor-out all fixtures into `conftest.py`

As good as it gets, holler if it breaks!

svenpeter42 commented:
Did a few quick tests today and everything's been working on macOS. Happy to merge this once I get confirmation that it also works on Linux from someone else

alxdrl commented:
@svenpeter42 was wondering about the new test workflow especially regarding incured costs for macOS jobs. It might be a concern as  it is 10x more expensive ? You be the judge.

asdfugil commented:
GitHub actions is free for public repositories.

https://docs.github.com/en/billing/managing-billing-for-your-products/managing-billing-for-github-actions/about-billing-for-github-actions#about-billing-for-github-actions

alxdrl commented:
> GitHub actions is free for public repositories.
> 
oh good to know, it was not the case two years ago ? I've been charged when building the kernel in CI on macOS and forgot to disable the workflow.

svenpeter42 commented:
Let's merge this and see if anyone runs into issues and fix them then!
